overlord/registrystate/registrystate.go

```diff
@@ -199,3 +212,27 @@ func updateDatabags(st *state.State, databag registry.JSONDataBag, reg *registry
 	st.Set("registry-databags", databags)
 	return nil
 }
+
+type cachedRegistryTx struct{}
+
+// RegistryTransaction returns the registry.Transaction cached in the context
+// or creates one and caches it, if none existed. The context must be locked by
+// the caller.
+func RegistryTransaction(ctx *hookstate.Context, reg *registry.Registry) (*registry.Transaction, error) {
+	tx, ok := ctx.Cached(cachedRegistryTx{}).(*registry.Transaction)
+	if ok {
+		return tx, nil
+	}
+
+	tx, err := newTransaction(ctx.State(), reg)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx.OnDone(func() error {
+		return tx.Commit()
```

Comment:
is Commit a no-op if we didn't change anything?

Reply:
Effectively, yes. It still writes but it writes what it reads basically. We can optimise it (there was also another transaction-related optimisation to do, but I've never gotten around to that ticket)

Reply:
we definitely want to look into that